keps/sig-storage/20190125-online-growing-persistent-volume-size.md

```diff
@@ -0,0 +1,139 @@
+
+### Risks and Mitigations
+
+- All in-tree volume plugins that support offline file system expansion support online file system expansion:
+  - GCE PD
+  - Azure Disk
+  - AWS EBS
+  - RBD
+  - Cinder
+
+However, Azure Disk in particular does not support online "device" expansion, so in practice a pod needs to be restarted and the device reattached for resize to happen regardless of whether this feature is enabled. File system expansion follows "device"/"controller"/"cloud provider" expansion (i.e. file system expansion is the resize2fs call to enlarge a file system to fill a device, and device expansion is the API call to expand the device in the first place). This difference is surfaced to the user via the "FileSystemResizePending" and "Resizing" conditions, for when file system or cloud provider expansion are in progress respectively. No additional conditions or configuration options will be surfaced to account for this case of online device expansion being unsupported: it is left to the volume plugin implementor & cloud provider to document it and return meaningful errors, which are displayed as events on PVCs.
+
+- Offline resize was introduced first and requires a pod restart to take effect, so online resize becoming the default and occurring immediately upon a PVC edit without requiring a restart may come as a surprise to users. But expansion of mounted file systems is supported by the kernel, so all applications will tolerate it. Plus, if there are cases where users would prefer their volumes to not resize immediately, they can still force offline resize to happen by stopping their pod before editing the PVC, then restarting it. Online resizing will be made default.
+
+- Enabling support of volume expansion through stateful sets (https://github.com/kubernetes/enhancements/issues/661) is orthogonal to making online resizing default. Stateful apps/workloads may need to restart to detect file systems' size changes but should have no issue with online resize.
+
+## Test Plan
+
+There will be e2e tests for online resizing. To isolate vs offline resize, the test will edit the PVC while the Pod is started instead of before.
+
+## Graduation Criteria
+
+### Alpha to Beta
+
+- e2e test where the PVC is edited while the Pod is started instead of before
+- update existing e2e tests so that they test offline resize correctly whether the online resize feature is enabled or not.
+- rename the operation metric to "online_node_expansion"
+
+### Beta to GA
+
+- Time for feedback (at least 1 release)
+- stress tests where multiple Pods are started/stopped & PVCs edited simultaneously to find race conditions in volume manager
+
+## Implementation History
+
+- 1.11: alpha
+- 1.15: beta
```

Comment:
oh sorry one more thing, I think we shouldn't add this here until it's actually done